nmdc_schema/migration_recursion.py

```diff
@@ -283,11 +292,114 @@ def __init__(self) -> None:
 
     def force_research_study_study_category(self, study: dict) -> dict:
         if 'study_category' not in study:
-            logger.info(f"Forcing 'study_category: research_study' on {study['id']}")
+            logger.info(
+                f"Forcing 'study_category: research_study' on {study['id']}")
             study['study_category'] = 'research_study'
         return study
 
 
+class Migrator_from_8_1_0_to_9_0_0(MigratorBase):
+    """Migrates data from schema 8.1.0 to 9.0.0"""
+
+    def __init__(self) -> None:
+        """Invokes parent constructor and populates collection-to-transformations map."""
+
+        super().__init__()
+
+        # Populate the "collection-to-transformers" map for this specific migration.
+        self.agenda = dict(
+            study_set=[self.fix_award_dois, self.fix_pub_dois,
+                       self.fix_massive, self.fix_ess_dive, self.remove_doi_slots],
+        )
+
+    def process_doi(self, study: dict, doi_list, doi_category):
+        id_value = study['id']
+        for doi_updates in doi_list:
+            if id_value == doi_updates['id']:
+                new_doi = {
+                    'doi_value': doi_updates['doi'],
+                    'doi_category': doi_category,
+                    'doi_provider': doi_updates['doi_prov']
+                }
+                study.setdefault('associated_dois', []).append(new_doi)
+
+    def fix_award_dois(self, study: dict):
+        """Moves current DOIs in the award_dois slot to the new associated_dois slot.
+         It also changes some DOIs that were incorrectly labeled as award to dataset DOIs.
+         It also add three new award DOIs from JGI"""
+
+        study_doi_data = self.load_yaml_file(
+            filename='assets/misc/study_dois_changes.yaml')
+
+        self.process_doi(
+            study, study_doi_data['award_move_to_data'], 'dataset_doi')
+        self.process_doi(study, study_doi_data['new_award_dois'], 'award_doi')
+        self.process_doi(study, study_doi_data['award_doi_prov'], 'award_doi')
+
+        return study
+
+    def fix_pub_dois(self, study: dict):
+        """Move publication_dois values to new associated_dois slot"""
+
+        study.setdefault('associated_dois', []).extend(
```

Comment:
Note to self: 

- `dictionary.setDefault('key', default)` ([docs](https://docs.python.org/3/library/stdtypes.html#dict.setdefault))
   > If `key` is in the dictionary, return its value. If not, insert `key` with a value of `default` and return `default`. `default` defaults to `None`.

This code is saying:
- given a dict named `study`
- if that dict already has a key named `associated_dois`, get the value of that key (presumably that value is a list); otherwise create that key and initialize it to an empty list
- given the list resulting from the previous step, append to it the values in the following list (which is constructed by the next two lines of code...)

Reply:
Note to self:

This code returns a [generator](https://realpython.com/introduction-to-python-generators/#using-generators) (I don't know where the topic is covered in the official Python docs)):
```py
{'doi_value': pub_doi, 'doi_category': 'publication_doi'}
            for pub_doi in study.get('publication_dois', [])
```

That generator is being passed as the `iterable` argument to `list.extend(iterable)`.

Reply:
The result of this statement is that `study.associated_dois` will contain a list of dictionaries, where each dictionary has the shape `{'doi_value': ..., 'doi_category': 'publication_doi'}`.